Convert translated R code to native Julia syntax

mitchphillipson opened the pull request:
This is a fairly substantial update with many new features. 

## Basic usage example

```julia
using HeaderArrayFile
const HAR = HeaderArrayFile # Create alias for convenience

using DataFrames

X = HAR.File("path/to/file.har") # Load HAR file

C = X["header_name"] # Access record by header name

DataFrame(C) # This is implemented for HarParameter types

HAR.internal_data(X) # List all internal data in the file

HAR.sets(X) # List all sets in the file
HAR.parameters(X) # List all parameters in the file

HAR.not_loaded(X) # List all records that are not loaded
```
# New features

1. The `not_loaded` function. There are several datatypes which neither of our versions support. This lets users see what data they aren't loading. 
2. I make a differentiation between `sets` and `parameters`, sets are `1C` data types and `parameters` are, essentially, everything else. 
3. The ability to make a DataFrame easily

# Before Merge

I currently am not returning these as arrays, I'm using DataFrames. I'd like to fix this before a new release so that people can choose different sinks. The goal is to implement the Tables.jl interface, which is the standard Julia method for this type of data.

mitchphillipson committed:
Updates to new structure

mitchphillipson committed:
Add convenience functions

mitchphillipson committed:
Implement output as a NamedArray for records

mitchphillipson committed:
Add docstrings on all functions

mitchphillipson committed:
Setting up Documenter

mitchphillipson committed:
Fix bug in retrieving indices

mitchphillipson committed:
Add tests

mitchphillipson committed:
Remove `Test` and `TestItemRunner` from compat

mitchphillipson committed:
Fix typo in tests

mitchphillipson committed:
Add testing data

mitchphillipson committed:
Add index.md

mitchphillipson committed:
Remove old documentation file

mitchphillipson committed:
Remove nightly from tests

mitchphillipson commented:
Added:

1. Support for `NamedArrays`, added example converting data to both a `NamedArray` and `DataFrame`
2. Added test to ensure data is loaded correctly
3. Added basic documentation

mivanic commented:
@mitchphillipson my apologies that it is taking me that long. I have finally taken a good look at your work.

Let me preface my comments with my appreciation for the rigorous and modern approach that you followed. I think this is the way to go.

Some issues that I wanted to bring to your attention:

(1) Single numbers are not supported and so in this example, you get an error

```
using HeaderArrayFile, NamedArrays
parameters = HeaderArrayFile.File("./gsdgpar.har")
NamedArray(parameters["dver"])
```

This is a problem because my GTAP-related work in Julia needs HAR files represented as they were intended in Fortran--an unordered dictionary of NamedArrays, e.g., doing something like:

```
parameters = Dict(
    k => NamedArray(parameters[k])
    for k ∈ collect(keys(parameters))
)
```

But this is no longer possible

(2) HAR files are not case sensitive because GEMAPCK is case insensitive, and therefore they are allowed to be internally inconsistent (famously the GTAP database may have capital spelled as "Capital" in one set and "capital" in another; similarly you can find Natres or natres or NatRes). I would recommend to either move all text to lowercase as I was doing in the current version (or allow for an optional switch to override it) because otherwise any user of HAR files in Julia would have to do this tedious work themselves.
[gsdgpar.zip](https://github.com/user-attachments/files/22753564/gsdgpar.zip)

mitchphillipson committed:
Add support for RL arrays and non-dimensional data

mitchphillipson committed:
Add support for RLFULL array

mitchphillipson committed:
Add support to normalize strings on load and optionally load unsupported types

mitchphillipson committed:
Update `show` methods

mitchphillipson commented:
## Point 1

The problem is a little more subtle than I initially expected. The "dver" variable is an RLFULL array and it's the only RLFULL array in the GTAP9 database. The real reason your extraction code was breaking is that any unsupported type still gets loaded, but because we don't know anything about it `NamedArray` fails. I'll add an option to not store unsupported types.

Here is code that is more stable:

```julia
using HeaderArrayFile, NamedArrays
const HAR = HeaderArrayFile

X = HeaderArrayFile.File("./gsdgpar.har")
NamedArray(X["dver"]) # This didn't work before because `dver` wasn't loaded

sets = Dict(key => NamedArray(value) for (key, value) in HAR.sets(X))
parameters = Dict(key => NamedArray(value) for (key, value) in HAR.parameters(X)) # This is the way to load the parameters. `1C` arrays are sets

HAR.not_loaded(X) # This is now empty. Before it had the data for `dver`.
```

If you know that `HAR.not_loaded(X)` is empty, this code will extract everything:

```julia
Dict(keys(X) .=> NamedArray.(values(X)))
```

Interestingly, neither your code nor the `gams.transfer` python package loaded the `dver` data. I believe the correct value should be 5. 

Also, I don't have very many examples of RL data, so this could be buggy on more general data. 

## Point 2

I've added two optional keywords to `HAR.File`:

- `normalizenames`: A function to normalize header names (default is `lowercase`).
- `load_unsupported`: A boolean indicating whether to load unsupported data types (default is `false`).

I've added a test to ensure the names are normalized.

## Point 3

I updated the `show` method for `HarFile` and `HarRecord` so that the data prints nicely.

mitchphillipson committed:
Update internal data check to lowercase

mivanic commented:
@mitchphillipson, thank you for your thoughtful responses--I think we are almost there! Also, please accept my apologies for being so slow in responding--I have had a very busy week. :(

The issue I ran into with the latest GTAP data is that Unicode in HAR files is not supported even though they started adding Unicode labels to the latest data. (I had to add it to my package after version 11 was released which already had Unicode characters--feel free to use my approach or develop your own.)

Here is the code that would not work:

```
sets_header = HeaderArrayFile.File("gsdfset.har")
```

[gsdfset.zip](https://github.com/user-attachments/files/22935104/gsdfset.zip)

mitchphillipson committed:
Update package name in README

mitchphillipson committed:
Add unicode decoding for unicode names

mitchphillipson commented:
Good catch! I've added a similar decoding step when loading sets.

Let me know if you find anything else.

mivanic commented:
Hi @mitchphillipson!

Great work! I managed to load the sets. 

I apologize for more issues though... :( I noticed that an integer header did not get loaded in the parameter file (and I do not see it as not loaded either). In the attached HAR file, it is header RDLT which is a type 2I. I have the code for it in my original package if you would like to have a look, but it is a simple matrix of integer, if I remember correctly. I am very hopefully this will be last issue before your changes can load a typical GTAP project.

[gsdfpar.zip](https://github.com/user-attachments/files/23106329/gsdfpar.zip)

mitchphillipson committed:
Add support for 2I and 2R types

mitchphillipson commented:
I'm working on this now, but I have a question. Before I answer, here is code to load the data with the unsupported 2I. I'm hoping the parameter not loading is due to `load_unsupported` being set to false.

```julia
using HeaderArrayFile
const HAR = HeaderArrayFile

X = HAR.File(
    raw"gsdfpar.har",
    normalizenames = lowercase,
    load_unsupported = true
)
```
Here is the actual question: Is your code loading 2IFULL correctly? This is what I see for the `rdlt` parameter

Julia code for replication:

```julia
P = X["rdlt"]

HAR.metadata(P)
D = HAR.data(P)
```

Metadata: 
    - Description - "Binary switch mechanism of allocating investment funds"
    - Dimensions - [1,1]

Assumingly, this means the resulting data is a singleton. The data `D` has a single row with 32 bytes, note that the four starting NULL bytes are excluded in the data. This corresponds to an array with eight 1's.
```julia
D.data[1] |> x -> reinterpret(Int32, x)
```
My interpretation of this is the first 7 `Int32`'s are metadata and any remaining are the actual data in that row. This roughly matches the formatting of a RESPSE array. 

However, I don't think this is what your code is doing so I am unsure of the correct output. Here is the excerpt from your code that I am following:

```julia
    elseif toRet["type"] == "2IFULL"
        combinedData = reduce((a, f) -> append!(f[17:length(f)], a), hh[2:length(hh)], init=Vector{UInt8}())
        combinedBuffer = IOBuffer(combinedData)
        combinedIntegers = map(f -> read(combinedBuffer, Int32), 1:(toRet["dimensions"][1]*toRet["dimensions"][2]))
        toRet["values"] = reshape(combinedIntegers, (toRet["dimensions"][1], toRet["dimensions"][2]))
```
In this example your `hh[2][5:end]` matches my `D.data`. The issue I see is that your `combinedIntegers` is going to grab only the first integer from `combinedData` (fifth from `hh[2]`), discarding the rest.

I received the following AI generated documentation from Florian Schiffmann concerning 2R and 2I arrays:

- Purpose: 2D arrays, float32 (2R) or int32 (2I).
- Second record specifics:
  - MaxDim = 2
  - dims = [nRows, nCols]
  - StorageType must be FULL.
- Data records (tiled blocks):
  - Tile header record:
    - "    ", nrec (int countdown), nRows_total, nCols_total,
      row_start, row_end, col_start, col_end (1‑based inclusive)
  - Tile payload record:
    - (row_end-row_start+1) × (col_end-col_start+1) elements
    - dtype: float32 for 2R, int32 for 2I
- Tiles collectively cover the matrix; read/write in Fortran order.

This matches my intuition, the `Data Records` tile starts with 4*8 bytes of metadata (NULL header included). 

Sorry for the novel, but I wanted to give you a full picture of my thoughts. I've pushed my interpretation so you can test. Let me know if I'm wrong and I'll correct the code. If you have more examples of `2I` data that would be ideal.

I've also added support for `2R` data types as the format is, essentially, identical to `2I`.

mivanic commented:
@mitchphillipson : It is possible that my code was indeed not working correctly and I only got lucky because I never had any issues with it. It is possible that I might have been reading the flags for fixed/mobile factors wrong, but I am sure I would have noticed that. At any rate, I tested your code and it works well. I have run a sample GTAP simulation and the Walras law holds, which means that, as far as I am concerned, the simulation worked at least as well as with my earlier version. Good job! Shall I merge?

mitchphillipson committed:
Update README

mitchphillipson commented:
Yes, I believe this should be good to go. Thanks for the help bug squashing.
